## Data Cleaning - handling missing values

Data cleaning is an important part of data science but it can be easily overlooked. Cleaning your data affects how your models behave and visuals appear. It greatly affects the insights you intend to discover. 

There are many data cleaning activities like missing values, parsing date, scalling and normalization. Starting with missing values, lets dig in.

### Imports

In [2]:
import pandas as pd
import numpy as np

## Load the data

In [3]:
df = pd.read_csv('new-york-history.csv')

# set seed for reproducibility
np.random.seed(0) 

In [4]:
df.head(6)

date state    death  deathConfirmed  deathIncrease  deathProbable  \
0  2021-03-01    NY  38577.0             NaN             80            NaN   
1  2021-02-28    NY  38497.0             NaN             90            NaN   
2  2021-02-27    NY  38407.0             NaN             86            NaN   
3  2021-02-26    NY  38321.0             NaN             94            NaN   
4  2021-02-25    NY  38227.0             NaN             92            NaN   
5  2021-02-24    NY  38135.0             NaN            104            NaN   

   hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
0       89995.0                 89995.0                 5307.0   
1       89995.0                 89995.0                 5259.0   
2       89995.0                 89995.0                 5445.0   
3       89995.0                 89995.0                 5626.0   
4       89995.0                 89995.0                 5703.0   
5       89995.0                 89995.0                 5876.0   

   hospitalizedIncrease  ...  totalTestResults  totalTestResultsIncrease  \
0                     0  ...          38281073                    174158   
1                     0  ...          38106915                    273720   
2                     0  ...          37833195                    285307   
3                     0  ...          37547888                    291189   
4                     0  ...          37256699                    278942   
5                     0  ...          36977757                    216813   

   totalTestsAntibody  totalTestsAntigen  totalTestsPeopleAntibody  \
0                 NaN                NaN                       NaN   
1                 NaN                NaN                       NaN   
2                 NaN                NaN                       NaN   
3                 NaN                NaN                       NaN   
4                 NaN                NaN                       NaN   
5                 NaN                NaN                       NaN   

   totalTestsPeopleAntigen  totalTestsPeopleViral  \
0                      NaN                    NaN   
1                      NaN                    NaN   
2                      NaN                    NaN   
3                      NaN                    NaN   
4                      NaN                    NaN   
5                      NaN                    NaN   

   totalTestsPeopleViralIncrease  totalTestsViral  totalTestsViralIncrease  
0                              0              NaN                        0  
1                              0              NaN                        0  
2                              0              NaN                        0  
3                              0              NaN                        0  
4                              0              NaN                        0  
5                              0              NaN                        0  

[6 rows x 41 columns]

In [5]:
df.shape

(365, 41)

The dataset has 365 rows and 41 columns. From the first six records displayed by `df.head()`, there are records with NaNs not a number value.NaNs are missing values that needs to be removed or replaced. Before we proceed to make that decision, let's determine the number of NaNs in each column.

In [6]:
df.isnull().sum()

date                                  0
state                                 0
death                                13
deathConfirmed                      364
deathIncrease                         0
deathProbable                       365
hospitalized                         19
hospitalizedCumulative               19
hospitalizedCurrently                15
hospitalizedIncrease                  0
inIcuCumulative                     365
inIcuCurrently                       24
negative                            365
negativeIncrease                      0
negativeTestsAntibody               365
negativeTestsPeopleAntibody         365
negativeTestsViral                  365
onVentilatorCumulative              365
onVentilatorCurrently                66
positive                              0
positiveCasesViral                  365
positiveIncrease                      0
positiveScore                         0
positiveTestsAntibody               365
positiveTestsAntigen                365


## Drop the columns with only NaN values

20 columns are empty as the have 365 null values in them. **deathConfirmed** has 364 NaNs meaning it is almost empty. Columns with only NaN values will be dropped followed by `'deathComfirmed'`. This will give us room to remove rows with NaNs using `dropna()` 

`df.dropna(axis=1)` could be used but this will mean droping columns with significance like `'hospitalizedCurrently'` and `'deathConfirmed'`

In [7]:
#counting columns that have Nas
df2 = df[[column for column in df if df[column].count()>=1]]
df2.shape

(365, 21)

In [8]:
df2.isnull().sum()

date                                  0
state                                 0
death                                13
deathConfirmed                      364
deathIncrease                         0
hospitalized                         19
hospitalizedCumulative               19
hospitalizedCurrently                15
hospitalizedIncrease                  0
inIcuCurrently                       24
negativeIncrease                      0
onVentilatorCurrently                66
positive                              0
positiveIncrease                      0
positiveScore                         0
totalTestEncountersViral              0
totalTestEncountersViralIncrease      0
totalTestResults                      0
totalTestResultsIncrease              0
totalTestsPeopleViralIncrease         0
totalTestsViralIncrease               0
dtype: int64

In [9]:
#removing deathconfirmed with just 1 value
df3 = pd.concat([df2.iloc[:, 0:3],df2.iloc[:,4:]],axis=1)
df3.shape

#this operation may be performed using
# df3 = df2.loc[:,['date','state', 'death', 'deathIncrease', 'hospitalized', 'hospitalizedCumulative', 'hospitalizedCurrently',
#        'hospitalizedIncrease', 'inIcuCurrently', 'negativeIncrease', 'onVentilatorCurrently', 'positive', 'positiveIncrease',
#        'positiveScore', 'totalTestEncountersViral', 'totalTestEncountersViralIncrease', 'totalTestResults',
#        'totalTestResultsIncrease', 'totalTestsPeopleViralIncrease','totalTestsViralIncrease']]

(365, 20)

In [10]:
#Remove rows with NaNs
df4 = df3.dropna()
df4.shape

(299, 20)

In [11]:
df4.head()

date state    death  deathIncrease  hospitalized  \
0  2021-03-01    NY  38577.0             80       89995.0   
1  2021-02-28    NY  38497.0             90       89995.0   
2  2021-02-27    NY  38407.0             86       89995.0   
3  2021-02-26    NY  38321.0             94       89995.0   
4  2021-02-25    NY  38227.0             92       89995.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                 89995.0                 5307.0                     0   
1                 89995.0                 5259.0                     0   
2                 89995.0                 5445.0                     0   
3                 89995.0                 5626.0                     0   
4                 89995.0                 5703.0                     0   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  positive  \
0          1065.0                 0                  741.0   1636680   
1          1083.0                 0                  728.0   1630445   
2          1121.0                 0                  753.0   1622865   
3          1132.0                 0                  771.0   1614724   
4          1124.0                 0                  774.0   1606520   

   positiveIncrease  positiveScore  totalTestEncountersViral  \
0              6235              0                  38281073   
1              7580              0                  38106915   
2              8141              0                  37833195   
3              8204              0                  37547888   
4              8746              0                  37256699   

   totalTestEncountersViralIncrease  totalTestResults  \
0                            174158          38281073   
1                            273720          38106915   
2                            285307          37833195   
3                            291189          37547888   
4                            278942          37256699   

   totalTestResultsIncrease  totalTestsPeopleViralIncrease  \
0                    174158                              0   
1                    273720                              0   
2                    285307                              0   
3                    291189                              0   
4                    278942                              0   

   totalTestsViralIncrease  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0

The dataset is void of missing values and now has 299 rows and 20 columns 

### Replace missing values 

It is possible to replace missing values using `fillna()` but this decision is relies on the intuition of the data analysts. The data analyst should try to understand why the data is missing. Is the data missing because it wasn't recorded or because it doesn't exist? If a value is missing because it doesn't exist, then it is best not to guess it. On the other hand, if a value is missing because it is not recorded, the the analyst can guess what it might be by looking at the other values in the column and row.

We may decide to replace NaNs in `'death'` with 0 because it is intuitive that noone die on that day and the the entry was omitted. Same can be said of `'hospitalized'`,`'inIcuCurrently'`,`'onVentilatorCurrently'`. If the analyst is are doing a meticulous analysis, he may go through every column to try and decide whoch value to use as a replacement. One must be cautious as this operation may end up adding noise to your data.

In [12]:
df5 = df2.loc[:,['date','state', 'death', 'deathIncrease', 'hospitalized', 'hospitalizedCumulative', 'hospitalizedCurrently',
       'hospitalizedIncrease', 'inIcuCurrently', 'negativeIncrease', 'onVentilatorCurrently', 'positive', 'positiveIncrease',
       'positiveScore', 'totalTestEncountersViral', 'totalTestEncountersViralIncrease', 'totalTestResults',
       'totalTestResultsIncrease', 'totalTestsPeopleViralIncrease','totalTestsViralIncrease']]

#replace NaNs with 0
df5[['death','hospitalized','hospitalizedCurrently','onVentilatorCurrently','inIcuCurrently']]=df5[['death','hospitalized','hospitalizedCurrently','onVentilatorCurrently','inIcuCurrently']].fillna(0)

In [13]:
df5.isnull().sum()

date                                 0
state                                0
death                                0
deathIncrease                        0
hospitalized                         0
hospitalizedCumulative              19
hospitalizedCurrently                0
hospitalizedIncrease                 0
inIcuCurrently                       0
negativeIncrease                     0
onVentilatorCurrently                0
positive                             0
positiveIncrease                     0
positiveScore                        0
totalTestEncountersViral             0
totalTestEncountersViralIncrease     0
totalTestResults                     0
totalTestResultsIncrease             0
totalTestsPeopleViralIncrease        0
totalTestsViralIncrease              0
dtype: int64

In [14]:
df6 = df5.dropna()
df6.shape

(346, 20)

By replacing the NaNs values with 0, we are now have a dataset of 346 rows and 20 columns. Other parameters for [`fillna()` can be found here](https://www.w3resource.com/pandas/dataframe/dataframe-fillna.php) 
